server/src/main/java/org/elasticsearch/action/admin/indices/datastream/GetDataStreamAction.java

```diff
@@ -169,22 +198,31 @@ protected void masterOperation(Task task, Request request, ClusterState state,
             Map<String, DataStream> dataStreams = clusterState.metadata().dataStreams();
 
             // return all data streams if no name was specified
-            final String requestedName = request.name == null ? "*" : request.name;
+            final String[] requestedNames = request.names == null || request.names.length == 0 ? new String[]{"*"} : request.names;
 
-            final List<DataStream> results = new ArrayList<>();
+            boolean hasWildcard = false;
+            final Set<DataStream> results = new HashSet<>();
+            for (String requestedName : requestedNames) {
                 if (Regex.isSimpleMatchPattern(requestedName)) {
```

Comment:
I think the wildcard logic is incorrect here with security enabled? This test fails: `org.elasticsearch.xpack.security.CoreWithSecurityClientYamlTestSuiteIT > test {yaml=indices.data_stream/10_basic/Get data stream}`

In this case there are no matching resources, so the expression gets rewritten to: `*,-*`. With this logic here,
this means that all data streams get included, because there is no logic that removes data streams if `-*` is used. We can add this logic (logic that deals with -[something] here, or we can use the index name expressions resolver and filter out anything that isn't a data stream?

Reply:
Good catch. I fixed the underlying name resolution issue here: https://github.com/elastic/elasticsearch/pull/59191